src/qibo/noise.py

```diff
@@ -93,16 +93,19 @@ def apply(self, circuit):
                 to the initial circuit with noise gates added according
                 to the noise model.
         """
-
-        circ = circuit.__class__(**circuit.init_kwargs)
+        noisy_circuit = circuit.__class__(**circuit.init_kwargs)
         for gate in circuit.queue:
-            circ.add(gate)
+            noisy_circuit.add(gate)
             if gate.__class__ in self.errors:
                 error, qubits = self.errors.get(gate.__class__)
                 if qubits is None:
                     qubits = gate.qubits
                 else:
                     qubits = tuple(set(gate.qubits) & set(qubits))
                 for q in qubits:
-                    circ.add(error.channel(q, *error.options))
-        return circ
\ No newline at end of file
+                    noisy_circuit.add(error.channel(q, *error.options))
+        noisy_circuit.parametrized_gates = list(circuit.parametrized_gates)
+        noisy_circuit.trainable_gates = list(circuit.trainable_gates)
```

Comment:
I think we should remove these two lines because:
1. In the original circuit these are not lists but [`_ParametrizedGates`](https://github.com/qiboteam/qibo/blob/20c6f3f22effbeccd5d31ed456717f9bee449e0c/src/qibo/abstractions/circuit.py#L13) objects. I am not sure if this makes a big difference but making them lists may break some other functionality (eg. using set_parameters, get_parameters in the noisy circuit).
2. Since we add all original gates properly in the noisy circuit using the `.add` method, this should properly update the `parametrized_gates` and `trainable_gates` lists and we don't need to worry about copying them.

Reply:
Ok, I'll remove them. For these lines I took inspiration from the [`with_noise`](https://github.com/qiboteam/qibo/blob/20c6f3f22effbeccd5d31ed456717f9bee449e0c/src/qibo/abstractions/circuit.py#L425-L426) method. I understand that since I am using `add` they will probably get updated automatically. 


> 1. In the original circuit these are not lists but [`_ParametrizedGates`](https://github.com/qiboteam/qibo/blob/20c6f3f22effbeccd5d31ed456717f9bee449e0c/src/qibo/abstractions/circuit.py#L13) objects. I am not sure if this makes a big difference but making them lists may break some other functionality (eg. using set_parameters, get_parameters in the noisy circuit).

Regarding this maybe it is better if we remove `list` in the corresponding lines of the `with_noise` method if you think that they can break some functionalities. Even if it is not strictly related to this PR I can do it here.

Reply:
> Regarding this maybe it is better if we remove `list` in the corresponding lines of the `with_noise` method if you think that they can break some functionalities. Even if it is not strictly related to this PR I can do it here.

You are right, this is not properly implemented in `with_noise` too. Here is an example:
```py
from qibo import gates
from qibo.models import Circuit

c = Circuit(2)
c.add(gates.RX(0, theta=0))
c.add(gates.U3(1, theta=0, phi=0, lam=0))
nc = c.with_noise((0.1, 0.1, 0.1))
```
If you do:
```py
c.set_parameters([0.1, 0.2, 0.3, 0.4])
```
it works, while
```py
nc.set_parameters([0.1, 0.2, 0.3, 0.4])
```
fails with an error that list has no attribute `nparams`.

If you like you can fix this, but remember to also change [`noisy_circuit.queue.append(gate)`](https://github.com/qiboteam/qibo/blob/20c6f3f22effbeccd5d31ed456717f9bee449e0c/src/qibo/abstractions/circuit.py#L422) to `noisy_circuit.add(gate)` so that all gates are properly added in the noisy circuit (as in your method). I see that there is a comment to not use `.add` because the gates were already added in the original circuit but I believe this is outdated now, as I tried switching this to `.add` and it works (all tests pass).

Reply:
Thanks for the example, now I see why this could be a problem. I can fix it now. 
I am also going to remove the comment since it is outdated.